include/boost/math/distributions/dirichlet.hpp

```diff
@@ -0,0 +1,405 @@
+} // mode
+
+
+template <class VectorType, class RealType, class Policy>
+inline RealType entropy(const dirichlet_distribution<VectorType, RealType, Policy> &dist)
+{
+  RealType t1 = 1;
+  for (size_t i = 0; i < dist.Order(); ++i)
+  {
+    t1 *= tgamma(dist.concentration[i]);
+  }
+  t1 = std::log(t1 / tgamma(dist.sum_concentration));
+  RealType t2 = (dist.sum_concentration - dist.Order()) * digamma(dist.sum_concentration);
+  RealType t3 = 0;
+  for (size_t i = 0; i < dist.Order(); ++i)
+  {
+    t3 += (dist.concentration[i] - 1) * digamma(dist.concentration[i]);
+  }
+  return t1 + t2 - t3;
+}
+
+
+template <class VectorType, class RealType, class Policy>
+inline RealType pdf(const dirichlet_distribution<VectorType, RealType, Policy> &dist, const VectorType &x)
+{ // Probability Density/Mass Function.
+  BOOST_FPU_EXCEPTION_GUARD
+
+  static const char *function = "boost::math::pdf(dirichlet_distribution<%1%> const&, %1%)";
+
+  BOOST_MATH_STD_USING // for ADL of std functions
+
+  // Argument checks:
+  RealType result = 0;
+  if (!dirichlet_detail::check_dist_and_x(function, x, &result, Policy()))
+  {
+    return result;
+  }
+  using boost::math::tgamma;
+  RealType f = 1;
+  for (size_t i = 0; i < dist.Order(); ++i)
+  {
+    f *= std::pow(x[i], dist.concentration[i] - 1);
+  }
+  f /= dist.normalizing_factor;
+  return f;
+} // pdf
+
+
+template <class VectorType, class RealType, class Policy>
+inline RealType cdf(const dirichlet_distribution<VectorType, RealType, Policy> &dist, const VectorType &x)
+{ // Cumulative Distribution Function dirichlet.
+  BOOST_MATH_STD_USING // for ADL of std functions
+
+      static const char *function = "boost::math::cdf(dirichlet_distribution<%1%> const&, %1%)";
+
+  // Argument checks:
+  RealType result = 0;
+  if (!dirichlet_detail::check_dist_and_x(function, dist.concentration, x, &result, Policy()))
+  {
+    return result;
+  }
+  RealType c = 1;
+  for (size_t i = 0; i < dist.Order(); ++i)
+  {
+    c *= std::pow(x[i], dist.concentration[i]) / tgamma(dist.concentration[i]) / dist.concentration[i];
+  }
+  c *= tgamma(dist.sum_concentration);
+  return c;
+} // dirichlet cdf
+
+template <class VectorType, class RealType, class Policy>
+inline RealType cdf(const complemented2_type<dirichlet_distribution<VectorType, RealType, Policy>, RealType> &c)
+{ // Complemented Cumulative Distribution Function dirichlet.
+
+  BOOST_MATH_STD_USING // for ADL of std functions
+
+  static const char *function = "boost::math::cdf(dirichlet_distribution<%1%> const&, %1%)";
+
+  RealType const &x = c.param;
+  dirichlet_distribution<RealType, Policy> const &dist = c.dist;
+
+  // Argument checks:
+  RealType result = 0;
+  if (!dirichlet_detail::check_dist_and_x(function, x, &result, Policy()))
+  {
+    return result;
+  }
+  RealType cumm = 1;
+  for (size_t i = 0; i < dist.Order(); ++i)
+  {
+    cumm *= std::pow(x[i], dist.concentration[i]) / tgamma(dist.concentration[i]) / dist.concentration[i];
```

Comment:
`std::pow` won't compile in `float128` precision. Use: `using std::pow` and then call `pow`. Also have a unit test showing compatibility with `#include <boost/multiprecision/float128.hpp>`.